service/service.go

```diff
@@ -342,24 +359,32 @@ func (app *Application) execute() {
 
 	app.asyncErrorChannel = make(chan error)
 
-	// Setup everything.
+	app.v.OnConfigChange(func(e fsnotify.Event) {
+		app.asyncErrorChannel <- errReloadConfig
+	})
+
 	app.setupTelemetry(ballastSizeBytes)
-	app.setupConfigurationComponents()
-	app.notifyPipelineReady()
+	for {
+		// Setup everything.
+		app.setupConfigurationComponents()
+		app.notifyPipelineReady()
 
-	// Everything is ready, now run until an event requiring shutdown happens.
-	app.runAndWaitForShutdownEvent()
+		// Everything is ready, now run until an event requiring shutdown happens.
+		err := app.runAndWaitForShutdownEvent()
 
-	// Begin shutdown sequence.
-	runtime.KeepAlive(ballast)
-	app.logger.Info("Starting shutdown...")
+		// Begin shutdown sequence.
+		runtime.KeepAlive(ballast)
+		app.logger.Info("Starting shutdown...")
 
-	app.notifyPipelineNotReady()
-	app.shutdownPipelines()
-	app.shutdownExtensions()
+		app.notifyPipelineNotReady()
+		app.shutdownPipelines()
+		app.shutdownExtensions()
 
+		if err != errReloadConfig {
+			break
+		}
```

Comment:
I'm not loving:
1. the distance between initializing and checking `err`,
1. the `for { ... if cond { break } }` pattern here.
🤔 could we `for err := errReloadConfig; err == errReloadConfig; { ... }`? I don't love that either 😞 . I also considered `return` here instead of `break`, but that'd require sticking the stuff after the loop into a `defer` func.

Reply:
I could break setup and teardown into different functions, then it might be a bit cleaner.

Reply:
is this cleaner?  Kind of annoying as you have to pass the ballast around.

Reply:
hrm I think I made it worse tbh.  reworking...